index.js

```diff
@@ -31,24 +33,98 @@ exports.register = (server, options) => {
-        // request scoped hub for breadcrumps and other scope metadata
-        request.__sentryDomain = domain.create();
-        request.__sentryDomain.enter();
-      }
+  // Set up request interceptor for creating and destroying a domain on each request
+  // It'll wrap the request handler returned by the _dispatch function in HAPI core
+  // allowing it to create the domain before any of HAPIs processing starts
+  // thus allowing us to use the normal HAPI extensions to add context to Sentry scopes
+  // https://github.com/hapijs/hapi/blob/c95985e225fa09c4b640a887ccb4be46dbe265bc/lib/core.js#L507-L538
+  function interceptor(next, req, res, ...args) {
+    const local = domain.create(); // Create domain to hold context for request
+    local.add(req);
+    local.add(res);
+
+    let rtn;
+
+    local.run(() => { // Create new scope for request
+      const currentHub = Sentry.getCurrentHub();
+
+      currentHub.configureScope(scope => {
+        scope.addEventProcessor(_sentryEvent => {
+          // format a sentry event from the request and triggered event
+          const sentryEvent = Sentry.Handlers.parseRequest(_sentryEvent, req);
+
+          // overwrite events request url if a baseUrl is provided
+          if (opts.baseUri) {
+            if (opts.baseUri.slice(-1) === '/') opts.baseUri = opts.baseUri.slice(0, -1);
+            sentryEvent.request.url = opts.baseUri + req.path;
+          }
+
+          // some SDK identifier
+          sentryEvent.sdk = { name: 'sentry.javascript.node.hapi', version };
+          return sentryEvent;
+        });
+      });
 
-      // attach a new scope to each request for breadcrumbs/tags/extras/etc capturing
-      request.sentryScope = new Sentry.Scope();
+      rtn = next.apply(this, [null, req, res].concat(args));
+    });
 
-      return h.continue;
-    },
+    return rtn;
+  }
+
+  // Wrap HAPI core _dispatch function. This function is primary entry point into HAPI for an
+  // external request. It's a factory that returns Node request handlers
+  // https://github.com/hapijs/hapi/blob/c95985e225fa09c4b640a887ccb4be46dbe265bc/lib/core.js#L505-L539
+  if (server._core && server._core._dispatch && !server._core._dispatch.__wrapped) {
+    shimmer.wrap(server._core, '_dispatch', function (original) { // eslint-disable-line
+      return function _dispatch_wrapped(...dispatchArgs) { // eslint-disable-line
+        const listener = original.apply(this, dispatchArgs);
+
+        function next(err, ...args) {
+          listener.apply(this, args);
+        }
+
+        return interceptor.bind(this, next);
+      };
+    });
+  }
+
+  // Setup listener interceptors so we can intercept inbound requests from Node, needed
+  // because we can't patch HAPI before it sets up listeners
+  eventsIntercept.patch(server.listener);
+  server.listener.intercept('request', function _requestInterceptor(...args) {
+    interceptor.apply(this, [args[args.length - 1], ...args.slice(0, -1)]);
```

Comment:
what is happening here lol

Reply:
Yeah this is bit mad. It's a cludge to make the `eventInterceptor` lib and `shimmy` lib accept the same function for request processing.

The `eventInterceptor` lib has a slightly nonsensical API, which results in this mess. Basically the `eventInterceptor` takes a function which will receive the intercepted event and has the signature `func(originalEventArg1, originalEventArg2, ..., originalEventArgN, next)` where `next` is the original function that should receive the event. Unfortunately, it's not possible to know how many arguments the original event will pass, so this bit of madness takes what `eventInterceptor` passes us and rearranges it so it matches this signature `func(next, originalEventArg1, ..., originalEventArgN)` which much easier to work with.

I suspect our best move here is to just replace `eventInterceptor` with our own thing that has a sane API. It's not a very complicated library.

Reply:
I've added a comment to explain what's happening. I think replacing the `eventInterceptor` is the right move, but should be done after the typescript conversion.

That way we can write a TS version of `eventInterceptor` that exposes the correct types on its interface, because currently there's a lot of `any`s floating around in my TS version to handle the lack of types on the `eventInterceptor` library.

Reply:
noted in #125